src/auth/mod.ts

```diff
@@ -13,35 +15,169 @@ interface DecodedToken {
-  return podsConfig.pods.some((pod) =>
-    pod.sub === payload.sub &&
-    pod.serviceaccount.uid === payload["kubernetes.io"]?.serviceaccount.uid
-  );
+export async function verifyServiceAccountToken(
+  token: string,
+): Promise<string> {
+  const payload = await verifyToken(token);
+
+  if (
+    !payload["kubernetes.io"] || !payload["kubernetes.io"].serviceaccount.name
+  ) {
+    const message = "Payload does not contain service account name field";
+    logger.error(message);
+    throw new HTTPException(HTTP_STATUS_CODES.UNAUTHORIZED, {
+      message,
+    });
+  }
+
+  return payload["kubernetes.io"].serviceaccount.name;
+}
+
+let jwks: JWK[] | undefined;
+const jwkExpiration = 24 * 60 * 60 * 1000; // 24 hours
+let expirationTime = Date.now() + jwkExpiration;
+
+function isJWKExpired(): boolean {
+  return Date.now() > expirationTime;
 }
 
-export function decodeToken(token: string): DecodedToken | null {
-  if (token.startsWith("Bearer ")) {
-    token = token.slice(7);
+const cryptoKeys: Map<string, CryptoKey> = new Map();
+async function verifyToken(token: string): Promise<DecodedToken> {
+  if (!jwks || isJWKExpired()) {
+    jwks = await getKeys();
   }
-  const [_headers, payload, _signature] = decode(token);
 
-  const data = payload as DecodedToken;
-  if (!data || typeof data !== "object" || !data.sub) {
-    logger.warn("Invalid token");
-    throw new HTTPException(401, {
-      message: "Invalid token",
-    });
+  const header = decode(token)[0];
+  const kid = (header as { kid?: string })?.kid;
+  if (!kid) {
+    const message = "Missing kid in token header";
+    logger.error(message);
+    throw new HTTPException(401, { message });
+  }
+  const matchingKey = jwks.find((k) => k.kid === kid);
+  if (!matchingKey) {
+    const message = "Key not found for kid in JWKs";
+    logger.error(message);
+    throw new HTTPException(401, { message });
+  }
+  const key = JSON.stringify(matchingKey);
+
+  const cryptoKey = cryptoKeys.has(kid)
+    ? cryptoKeys.get(kid)!
+    : await crypto.subtle.importKey(
+      "jwk",
+      JSON.parse(key),
+      { name: "RSASSA-PKCS1-v1_5", hash: "SHA-256" },
+      true,
+      ["verify"],
+    );
+
+  if (!cryptoKeys.has(kid)) {
+    cryptoKeys.set(kid, cryptoKey);
+  }
+
+  const verified = await verify(token, cryptoKey);
+  return verified as DecodedToken;
+}
+
+async function getKeys(): Promise<JWK[]> {
+  const jwks_uri = await getJWKURI();
+  const jwkUrl = new URL(jwks_uri);
+  const headers = new Headers();
+  if (envVarsConfig.env === "DEV") {
+    // This should be the proxy server created by kubectl for dev purposes inside a local machine
+    jwkUrl.hostname = envVarsConfig.k8s_api;
+  } else {
+    // we need to set the auth headers with the serviceToken when herald is running in a k8s cluster
+    headers.set(
+      "Authorization",
+      `Bearer ${await getServiceAccountToken()}`,
+    );
+  }
+  const fetchFunc = async () => await fetch(jwkUrl.toString(), { headers });
```

Comment:
I did not catch that, but this is not a timeout as per last comment, this is a retry. What happens if the connection never return anything but is left open? See the abortcontroller